compiler/src/iree/compiler/Dialect/Flow/Transforms/DynamicizeStaticShapes.cpp

```diff
@@ -0,0 +1,278 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <memory>
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "llvm/ADT/SetVector.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/FunctionInterfaces.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#define DEBUG_TYPE "iree-flow-dynamicize-static-shapes"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+struct DynamicizeStaticShapesPass
+    : public DynamicizeStaticShapesBase<DynamicizeStaticShapesPass> {
+  void runOnOperation() override;
+};
+
+/// Returns the corresponding dynamic shaped type of the given static shaped
+/// `type`.
+RankedTensorType getDynamicTensorType(RankedTensorType type) {
+  SmallVector<int64_t, 4> shape;
+  shape.resize(type.getRank(), ShapedType::kDynamic);
+  return RankedTensorType::get(shape, type.getElementType(),
+                               type.getEncoding());
+}
+
+/// Goes through the given list of types, return dynamic shaped counterparts for
+/// tensor shapes, and return non-tensor types as-is.
+SmallVector<Type> getDynamicTensorTypes(Operation::result_type_range types) {
+  SmallVector<Type> resultTypes;
+  for (Type type : types) {
+    if (auto tensorType = type.dyn_cast<RankedTensorType>()) {
+      resultTypes.push_back(getDynamicTensorType(tensorType));
+    } else {
+      resultTypes.push_back(type);
+    }
+  }
+  return resultTypes;
+}
+
+/// Returns a vector of SSA values representing each dimension of the given
+/// `value`'s shape.
+SmallVector<Value, 4> getDimValues(OpBuilder &builder, Location loc,
+                                   Value value, ValueRange dynamicDims) {
+  auto type = value.getType().cast<ShapedType>();
+  SmallVector<Value, 4> dimValues;
+  dimValues.reserve(type.getRank());
+
+  int dimIndex = 0;
+  for (int64_t dimSize : type.getShape()) {
```

Comment:
You can use https://github.com/llvm/llvm-project/blob/1def3141135c072a1d3e51e82e113dd67b0def97/mlir/include/mlir/Dialect/Tensor/Utils/Utils.h#LL36C55-L36C55 instead.